### PR TITLE
[GPU] Enable Metal on macosx

### PR DESCRIPTION
**Pull request by xta0**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48008 [Person Seg] Compress the person seg model
* **#47635 [GPU] Enable Metal on macosx**

Add macosx support for metal. The supported os version is 10.13 and above.

Differential Revision: [D24825088](https://our.internmc.facebook.com/intern/diff/D24825088/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24825088/)!

### CODE REVIEW

File: cmake/Metal.cmake [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    if(_CMAKE_OSX_SDKS) 
+        list(SORT _CMAKE_OSX_SDKS)
+        list(REVERSE _CMAKE_OSX_SDKS)
+        list(GET _CMAKE_OSX_SDKS 0 CMAKE_OSX_SDK_ROOT)
+        message(STATUS "_CMAKE_OSX_SDKS: ${_CMAKE_OSX_SDKS}")
+    else(_CMAKE_OSX_SDKS)
+        message(FATAL_ERROR "No OSX SDK's found in default search path ${CMAKE_OSX_DEVELOPER_ROOT}.")
+    endif(_CMAKE_OSX_SDKS)
+    message(STATUS "Toolchain using default OSX SDK: ${CMAKE_OSX_SDK_ROOT}")
+endif(NOT DEFINED CMAKE_OSX_SDK_ROOT)
```
Comment (ljk53) [inline cmake/Metal.cmake:33]: Seems very similar to iOS.cmake - any chance to factor out?
also cc: @malfet @xuhdev

File: aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp [hunk 2]
```diff
@@ -3,7 +3,7 @@
 #include <ATen/ATen.h>
 
 
-#if defined(C10_IOS)
+#if (C10_IOS || TARGET_OS_MAC)
```
Comment (ljk53) [inline aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp:6]: Nothing wrong about this - just feel that having C10_IOS (defined by us) and TARGET_OS_MAC (defined by apple) on the same condition is a little unbalanced, lol.

If this condition is going to be used at other places, shall we create a dedicated macro? Not sure what's the best convention in PyTorch codebase (cc: @dzhulgakov):
- if it's targeting apple devices, shall we introduce 'C10_APPLE'?
- or, can we simply use some master feature switch? e.g. '#if AT_METAL_ENABLED()' seems to be one of the common pattern.

File: aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp [hunk 3]
```diff
@@ -94,19 +94,19 @@ c10::intrusive_ptr<Conv2dOpContext> conv2d_prepack(
 Tensor conv2d_prepack_run(
     const Tensor& input,
     const c10::intrusive_ptr<Conv2dOpContext>& op_context) {
-#if defined(C10_IOS)
+#if (C10_IOS || TARGET_OS_MAC)
   return mpscnn::conv2d(input, *op_context);
 #else
-  TORCH_CHECK(false, "conv2d_prepack_run can only be invoked on iOS");
+  TORCH_CHECK(false, "conv2d_prepack_run can only be invoked on iOS and MacOS");
   return input;
```
Comment (ljk53) [inline aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp:101]: nit: I don't think we need 'return input;' here as TORCH_CHECK(false, ...) throws.

File: aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp [hunk 4]
```diff
@@ -94,19 +94,19 @@ c10::intrusive_ptr<Conv2dOpContext> conv2d_prepack(
...
 }
 
 Tensor copy_to_host(const Tensor& input) {
-#if defined(C10_IOS)
+#if (C10_IOS || TARGET_OS_MAC)
   return mpscnn::copy_to_host(input);
 #else
-  TORCH_CHECK(false, "copy_to_host can only be invoked on iOS");
+  TORCH_CHECK(false, "copy_to_host can only be invoked on iOS and MacOS");
   return input;
```
Comment (ljk53) [inline aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp:110]: ditto

File: caffe2/CMakeLists.txt [hunk 5]
```diff
@@ -1195,6 +1195,18 @@ if(USE_CUDA)
 
 endif()
 
+# ---[ Metal(OSX) modification
+if(APPLE AND USE_PYTORCH_METAL)
```
Comment (ljk53) [inline caffe2/CMakeLists.txt:1199]: nit: can we make 'USE_PYTORCH_METAL' only enabled when 'APPLE' is true?

**Review by ljk53 — COMMENTED** [review]
LGTM, just some nits and questions about convention.

File: aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp (hunk 2 above)
Comment (xta0) [reply aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp:6]: How about (TARGET_OS_PHONE || TARGET_OS_MAC). Would that make you feel better 🤣

File: caffe2/CMakeLists.txt (hunk 5 above)
Comment (xta0) [reply caffe2/CMakeLists.txt:1199]: `USE_PYTORCH_METAL` is also available in Linux. In that case, only one cpp file under `aten/native/metal` will be compiled. The reason is that users can still do `optimize_for_mobile` on Linux machines to export models for metal.

**Review by ljk53 — APPROVED** [review]
accept to unblock - we could clean up these bool conditions (C10_IOS || TARGET_OS_MAC) later if they are going to be used a lot in the codebase in the future.

**Comment by PeterL1n** [comment]
Does this mean PyTorch will support inferencing and training on Mac using GPU in the future?

**Comment by leovinus2001** [comment]
> Does this mean PyTorch will support inferencing and training on Mac using GPU in the future?

I am wondering the same. Could someone in the know write a few sentences pls as to the goal here? 
Train vs inference?
macOS vs iOS?
Metal throughout PyTorch or just ATen (base don checkins I see :)
Metal just for CNN or also Linear/LSTM?

When more clarity is provided then it is easier to see whether we can contribute :)

**Pull request closed without merging**